streams/src/main/java/org/apache/kafka/streams/state/internals/NamedCache.java

```diff
@@ -0,0 +1,306 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.kafka.streams.state.internals;
+
+import org.apache.kafka.common.metrics.Sensor;
+import org.apache.kafka.common.utils.Bytes;
+import org.apache.kafka.streams.KeyValue;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.TreeSet;
+
+class NamedCache {
+    private static final Logger log = LoggerFactory.getLogger(NamedCache.class);
+    private final String name;
+    private final TreeMap<Bytes, LRUNode> cache = new TreeMap<>();
+    private final Set<Bytes> dirtyKeys = new LinkedHashSet<>();
+    private ThreadCache.DirtyEntryFlushListener listener;
+    private LRUNode tail;
+    private LRUNode head;
+    private long currentSizeBytes;
+    private ThreadCacheMetrics metrics;
+
+    // JMX stats
+    private Sensor hits = null;
+    private Sensor misses = null;
+    private Sensor overwrites = null;
+    private Sensor flushes = null;
+
+    // internal stats
+    private long numReadHits = 0;
+    private long numReadMisses = 0;
+    private long numOverwrites = 0;
+    private long numFlushes = 0;
+
+    NamedCache(final String name) {
+        this(name, null);
+    }
+
+    NamedCache(final String name, final ThreadCacheMetrics metrics) {
+        this.name = name;
+        this.metrics = metrics != null ? metrics : new ThreadCache.NullThreadCacheMetrics();
+
+        this.hits = this.metrics.addCountSensor(name, "hits");
+        this.misses = this.metrics.addCountSensor(name, "misses");
+        this.overwrites = this.metrics.addCountSensor(name, "overwrites");
+        this.flushes = this.metrics.addCountSensor(name, "flushes");
+    }
+
+    synchronized long hits() {
+        return numReadHits;
+    }
+
+    synchronized long misses() {
+        return numReadMisses;
+    }
+
+    synchronized long overwrites() {
+        return numOverwrites;
+    }
+
+    synchronized long flushes() {
+        return numFlushes;
+    }
+
+    synchronized LRUCacheEntry get(final Bytes key) {
+        final LRUNode node = getInternal(key);
+        if (node == null) {
+            return null;
+        }
+        updateLRU(node);
+        return node.entry;
+    }
+
+    synchronized void setListener(final ThreadCache.DirtyEntryFlushListener listener) {
+        this.listener = listener;
+    }
+
+    synchronized void flush() {
+        metrics.recordCount(flushes, 1);
+        numFlushes++;
+
+        log.debug("Named cache {} stats on flush: #hits={}, #misses={}, #overwrites={}, #flushes={}",
+            name, hits(), misses(), overwrites(), flushes());
+
+        if (listener == null) {
```

Comment:
Will it be doomed to through this exception if user call `enableCaching` when creating the store by their own with `Stores`?